x/mongo/driver/topology/server.go

```diff
@@ -795,12 +795,15 @@ func (s *Server) check() (description.Server, error) {
 	var duration time.Duration
 
 	start := time.Now()
+
+	// Create a new connection if this is the first check, the connection was closed after an error during the previous
+	// check, or the previous check was cancelled.
 	if s.conn == nil || s.conn.closed() || s.checkWasCancelled() {
-		// Create a new connection if this is the first check, the connection was closed after an error during the previous
-		// check, or the previous check was cancelled.
+		connID := "0"
 		if s.conn != nil {
-			s.publishServerHeartbeatStartedEvent(s.conn.ID(), false)
+			connID = s.conn.ID()
 		}
+		s.publishServerHeartbeatStartedEvent(connID, false)
```

Comment:
We should do something similar for the subsequent `publishServerHeartbeatSucceededEvent` and `publishServerHeartbeatFailedEvent` so they always get published, even if the conn is nil. That's especially important for the failed event so we can report the error.